lib/obj.js

```diff
@@ -1,365 +1,447 @@
+var vertexPattern = /v( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                      // v float float float
+var normalPattern = /vn( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                     // vn float float float
+var uvPattern = /vt( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                                               // vt float float
+var facePattern1 = /f( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)?\/?/;                                                                        // f vertex vertex vertex ...
+var facePattern2 = /f( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)?/;                            // f vertex/uv vertex/uv vertex/uv ...
+var facePattern3 = /f( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))?/;    // f vertex/uv/normal vertex/uv/normal vertex/uv/normal ...
+var facePattern4 = /f( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))?/;                                // f vertex//normal vertex//normal vertex//normal ...
+
+/**
+ * Parse an obj file.
+ *
+ * @param {String} objPath Path to the obj file.
+ * @returns {Promise} A promise resolving to the obj data.
+ * @exception {RuntimeError} The file does not have any geometry information in it.
+ *
+ * @private
+ */
+function loadObj(objPath) {
+    // Global store of vertex attributes listed in the obj file
+    var positions = new ArrayStorage(ComponentDatatype.FLOAT);
+    var normals = new ArrayStorage(ComponentDatatype.FLOAT);
+    var uvs = new ArrayStorage(ComponentDatatype.FLOAT);
+
+    // The current node, mesh, and primitive
+    var node;
+    var mesh;
+    var primitive;
+
+    // All nodes seen in the obj
+    var nodes = [];
+
+    // Used to build the indices. The vertex cache is unique to each mesh.
+    var vertexCache = {};
+    var vertexCacheLimit = 1000000;
+    var vertexCacheCount = 0;
+    var vertexCount = 0;
+
+    // All mtl paths seen in the obj
+    var mtlPaths = [];
+
+    function getName(name) {
+        return (name === '' ? undefined : name);
+    }
+
+    function addNode(name) {
+        node = new Node();
+        node.name = getName(name);
+        nodes.push(node);
+        addMesh();
+    }
+
+    function addMesh(name) {
+        mesh = new Mesh();
+        mesh.name = getName(name);
+        node.meshes.push(mesh);
+        addPrimitive();
+
+        // Clear the vertex cache for each new mesh
+        vertexCache = {};
+        vertexCacheCount = 0;
+        vertexCount = 0;
+    }
+
+    function addPrimitive() {
+        primitive = new Primitive();
+        mesh.primitives.push(primitive);
+    }
+
+    function useMaterial(name) {
+        // Look to see if this material has already been used by a primitive in the mesh
```

Comment:
Yeah I could have done a material -> primitive dictionary. Since nodes and meshes are also arrays I just wanted to keep it consistent.